File: Pod/Classes/HSM/CPMessagesQueue.m
//
//  CPMessagesQueue.m
//  GCD的其他用法
//
//  Created by cqh on 15/1/27.
//  Copyright (c) 2015年 Cnepay. All rights reserved.
//

#import "CPMessagesQueue.h"
#import "CPMessageOperation.h"

@interface CPMessagesQueue ()
@property (nonatomic ,strong) CPMessageOperation *lastOperation;
@property (readwrite, nonatomic, strong) CPMessageOperation *excutingOperation;
@end

@implementation CPMessagesQueue
- (instancetype)init
{
    self = [super init];
    if (self) {
        self.maxConcurrentOperationCount = 1;
    }
    return self;
}

- (void)addOperationAtEndOfQueue:(CPMessage *)message {
    CPMessageOperation *op = [CPMessageOperation sendMessageOperationWithTask:message];
    op.runloopThread = self.runloopThread;
    op.delegate = self;
    
    if (self.lastOperation) {
        [op addDependency:self.lastOperation];
    }
    @synchronized (self) {
        self.lastOperation = op;
    }
    [super addOperation:op];
}


- (void)addOperationAtFrontOfQueue:(CPMessage *)message {
    CPMessageOperation *op = [CPMessageOperation sendMessageOperationWithTask:message];
    op.runloopThread = self.runloopThread;
    op.delegate = self;
    
    NSArray *operations = self.operations;
    [operations enumerateObjectsUsingBlock:^(CPMessageOperation* operation, NSUInteger idx, BOOL *stop) {
        if([operation isExecuting]){
            [op addDependency:operation];
        }else{
            [operation addDependency:op];
        }
    }];
    [super addOperation:op];
}


- (void)removeOperationWithMessage:(CPMessage *)message {
    
    NSArray *operations = self.operations;
    [operations enumerateObjectsUsingBlock:^(CPMessageOperation* operation, NSUInteger idx, BOOL *stop){
        if(message.type == operation.message.type) {
            [operation cancel];
        }
    }];
}

- (void)delayOperationAtEndOfQueue:(CPMessage *)message delay:(NSTimeInterval)delay {
    
    CPMessageOperation *op = [CPMessageOperation sendMessageOperationWithTask:message];
    op.runloopThread = self.runloopThread;
    op.delegate = self;
    op.delayTime = delay;
    if (self.lastOperation) {
        [op addDependency:self.lastOperation];
    }
    @synchronized (self) {
        self.lastOperation = op;
    }
    [super addOperation:op];
}


- (void)sendMessageOperation:(CPMessageOperation *)operation didFinishSendMessage:(CPMessage *)message {}
@end

File: Pod/Classes/HSM/CPMessagesQueue.swift
//
//  CPMessagesQueue.swift
//  StateMachine-Swift
//
//  Created by Cnepay on 15/10/18.
//  Copyright © 2015年 Cnepay. All rights reserved.
//

import UIKit


class CPMessagesQueue: NSOperationQueue {
    
    var lastOperation:CPMessageOperation?
    
    override init() {
        super.init()
        self.maxConcurrentOperationCount = 1;
    }
    
    func addOperationAtEndOfQueue(message:CPMessage) {
        let messageOperation = CPMessageOperation(message: message);
        messageOperation.delegate = self;
        
        if self.lastOperation != nil {
            messageOperation.addDependency(lastOperation!);
        }
        self.lastOperation = messageOperation;
        
        super.addOperation(messageOperation);
    }
    
    func addOperationAtFrontOfQueue(message:CPMessage) {
        let messageOperation = CPMessageOperation(message: message);
        messageOperation.delegate = self;
        
        for singleOperation in self.operations as! [CPMessageOperation] {
            if singleOperation.executing {
                messageOperation.addDependency(singleOperation);
            }else {
                singleOperation.addDependency(messageOperation);
            }
        }
        super.addOperation(messageOperation);
    }
    
    func removeOperationWithMessage(message:CPMessage) {
        for singleOperation in self.operations as! [CPMessageOperation] {
            if message.type ==  singleOperation.message.type {
                singleOperation.cancel()
            }
        }
    }
    
    
    
    //一个类型已经实现了协议中的所有要求，却没有声明为遵循该协议时，可以通过扩展(空的扩展体)来补充协议声明
    func sendMessageOperationDidStart(operation: CPMessageOperation, message: CPMessage) {
        
    }
    
}

extension CPMessagesQueue:CPMessageOperationDelegate {}
